README.md
# Gigabyte GA-Z77X DSDT Patch

Support files for running macOS 10.11+ on Gigabyte 7 Series motherboards.

This is a WIP branch for macOS Sierra. Support for other boards will be re-added soon.

These support files can help you configure the post-installation process on Gigabyte 7 Series motherboards. The current repository uses custom SSDTs for each motherboard along with the GA-Z77X.sh post-installation script to create a working macOS installation.

### Usage

```
git clone --recursive https://github.com/theracermaster/Gigabyte-GA-Z77X-DSDT-Patch --branch sierra
```
```
cd Gigabyte-GA-Z77X-DSDT-Patch
```
```
./GA-Z77X.sh --install
```

### Supported Motherboards
Currently, the following motherboards are supported by the GA-Z77X.sh script:
 - GA-Z77X-UD5H (-WB WIFI)
 - GA-Z77X-UP5 TH

### Issue Reporting
If you run into an issue while running the GA-Z77X.sh post-installation script or have an issue with the custom SSDTs, please report it on [GitHub Issues](https://github.com/theracermaster/Gigabyte-GA-Z77X-DSDT-Patch/issues).

### Credits
Thanks to:
 - [Mieze](https://github.com/Mieze) - [AtherosE2200Ethernet](https://github.com/Mieze/AtherosE2200Ethernet), [IntelMausiEthernet](https://github.com/Mieze/IntelMausiEthernet)
 - [netkas](http://netkas.org/) & [kozlek](https://github.com/kozlek) - [FakeSMC/HWSensors](https://github.com/RehabMan/OS-X-FakeSMC-kozlek)
 - [Pike R. Alpha](https://github.com/Piker-Alpha) - [ACPI table examples](https://github.com/Piker-Alpha/DSDT)
 - [RehabMan](https://github.com/RehabMan) - [Example SSDTs](https://github.com/RehabMan/OS-X-Clover-Laptop-Config)
 - [Shailua](http://www.insanelymac.com/forum/user/933246-shailua/) - [AtherosL1cEthernet](http://www.insanelymac.com/forum/topic/283086-updated-atheros-ar8131325152-driver-for-107108/)
 - [the-darkvoid](https://github.com/the-darkvoid) - [XPS9530-OSX Repo/Script](https://github.com/the-darkvoid/XPS9530-OSX)
 - [vit9696](https://github.com/vit9696) - [AppleALC](https://github.com/vit9696/AppleALC) & [Shiki](https://github.com/vit9696/Shiki)


CHANGELOG.md
# Gigabyte GA-Z77X-DSDT Patch Changelog

Version 2.0b1 (2016-10-15)
 - Added support for macOS 10.12 Sierra - removed support for 10.10 and lower
 - Refactored and rewrote parts of script
 - Refactored SSDTs into "generic" SSDTs & board-specific SSDTs
 - Replaced simpleMacSerial.sh/simpleMLB.sh with MacGen
 - Removed hdaInjector.sh - replaced with AppleALC.kext by vit9696
 - Removed ssdtPRGen.sh - replaced with Clover's C-state/P-state generation
 - Updated Clover to r3816-RM-29e557b
 - Added FireWire EFI drivers from iMac17,1 firmware
 - Added EfiDevicePathPropertyDatabase EFI driver from iMac17,1 firmware
 - Updated HfsPlus EFI driver from iMac17,1 firmware
 - Updated AtherosE2200Ethernet.kext & IntelMausiEthernet.kext by Mieze to 2.2.0
 - Added Shiki.kext by vit9696
 - Removed FakePCIID & GA-Z77X_USB.kext - replaced with Clover kext patches

Version 1.8.4 (2015-10-24)
 - Optimized & rewrote many methods in the GA-Z77X.sh script & fixed various minor bugs
 - Removed `dspci` (no longer needed)
 - Updated Clover to r3302 (built with GCC 5)
 - Corrected max turbo frequency detection during CPU PM SSDT generation
 - Corrected XCPM configuration (added checking to only enable XCPM on Ivy Bridge CPUs)
 - Fixed EFI system partition mounting on Core Storage volumes (thanks to mickelsn on GitHub Issues for reporting!)

Version 1.8.3 (2015-10-20)
 - Fixed GA-Z77X-UP5 TH SSDT (thanks to ch0w7 on GitHub Issues for reporting!)
 - Finally fixed PCIe GPU detection, kernel panics when using IGPU only are now fixed
 - Updated Clover to r3295 (built with GCC 5)

Version 1.8.2 (2015-10-12)
 - Updated Clover to custom build of r3280: this version injects kexts from kexts/Other & kexts/10.11 at the same time
 - Updated Mieze's IntelMausiEthernet driver to 2.0.2.d5
 - Removed ssdtPRGen.sh as a submodule, as it is downloaded instead
 - Updated SSDTs

Version 1.8.1 (2015-10-01)
 - Fixed FakeSMC.kext+HWSensors (HWSensors didn't load due to version errors)
 - Fixed config.plist failing to rename when HDA injection wasn't used (thanks to iHedgehog on GitHub Issues for reporting!)

Version 1.8.0 (2015-09-30)
 - Fixed GA-Z77X-UP5 TH motherboard detection (thanks to vodkaflip on GitHub Issues for reporting!)
 - Updated FakeSMC.kext+HWSensors with Maxwell GPU monitoring support
 - Added SSDT generation using the beta version of ssdtPRGen.sh by Piker-Alpha, as Clover SSDT generation didn't always work properly
 - Added experimental support for the GA-B75M-D3H motherboard

Version 1.7.9 (2015-09-25)
 - Added GA-Z77X_USB.kext for USB port injection on OS X 10.11 (thanks to RehabMan, heavily based off his work)
 - Fixed AppleUSBXHCI kext patches on OS X 10.9 (note: no OS X 10.11 AppleUSBXHCI patches yet)
 - Fixed FVInjector.kext on OS X 10.9/10.10

Version 1.7.8 (2015-09-05)
 - Updated iasl
 - Updated Mieze's RealtekRTL8111 driver (compiled from RehabMan's fork)
 - Moved COPR → MATH device renaming to Clover config.plist DSDT binpatch
 - Disabled Clover AppleRTC patch
 - Simplified non-Intel AppleUSBXHCI USB 3.0 controller kext patches
 - Added experimental support for the GA-Z77X-UD4H motherboard
 - Removed VoodooPS2Daemon (not needed on desktops)

Version 1.7.7 (2015-09-04)
 - Updated SSDTs with IGPU primary & discrete GPU secondary support
 - Updated SSDTs with `device_type` property injection for discrete GPU device (GFX0) & HDMI audio device (HDAU)
 - ACPI OS is now simulated as `Windows 2012` (Windows 8) with SSDT XOSI method & Clover DSDT binpatch (thanks to RehabMan)
 - Disabled Clover's USB injection since it's already done in the SSDTs

Version 1.7.6 (2015-09-04)
 - Renamed USB devices (EHCx, XHCx → EH0x, XH0x) using config.plist DSDT patches to bypass Apple's port restrictions
 - Removed GA-Z77X_USBPortInjector.kext
 - Fixed non-Intel XHCI controller kext patching
 - Fixed FakePCIID.kext+FakePCIID_XHCIMux.kext installation
 - Fixed various minor bugs in the script

Version 1.7.5 (2015-08-28)
 - Updated Mieze's IntelMausiEthernet driver
 - Updated Clover to r3262
 - Replaced AppleUSBEHCIPortInjector.kext & AppleUSBXHCIPortInjector.kext with GA-Z77X_USBPortInjector.kext
 - Added FreqVectorInjector.kext to suppress `X86PlatformShim::start - Failed to send stepper` errors in kernel logs

Version 1.7.4 (2015-08-24)
 - Refined motherboard detection (uses OEMBoard property injected by Clover in `IODeviceTree` instead of `bdmesg`)
 - Added AppleUSBXHCIPortInjector.kext creation for OS X 10.11 (experimental)
 - Fixed GA-Z77X-UP5 TH & GA-Z77X-UP7 motherboard detection
 - Slight code refactoring

Version 1.7.3 (2015-08-23)
 - Added Clover DSDT patches to generic config to make SSDTs smaller

Version 1.7.2 (2015-08-22)
 - Fixed more bugs reported by sreda2013 (thanks again for your patience)
 - Fixed more typos in the script
 - Added AppleUSBEHCIPortInjector.kext creation for OS X 10.11

Version 1.7.1 (2015-08-19)
 - Fixed more bugs related to kext installation & iasl (thanks to sreda2013 again for reporting them)
 - Fixed more typos in the script

Version 1.7 (2015-08-18)
 - Fixed motherboard detection bugs (thanks to sreda2013 for reporting them)
 - Added styling to main script
 - Added detection of other UEFI bootloaders on the ESP
 - Fixed bugs related to simpleMacSerial & simpleMLB
 - Fixed typos in script

Version 1.5 (2015-08-17)

tools/common.sh
#!/usr/bin/env zsh
#
# common.sh - Common functions for macOS post-installation scripts
# By theracermaster - https://github.com/theracermaster

## The user ID
gID=$(id -u)

## The major version of macOS
gOSVersion=$(sw_vers -productVersion | awk -F '.' '{print $1 "." $2}')

## The location where the EFI partition is mounted, will be properly initialized later
gEFIMount="Unknown"

## iasl binary
IASL="$gRepo/tools/iasl"

## Styling
STYLE_RESET="\e[0m"
STYLE_BOLD="\e[1m"
STYLE_UNDERLINED="\e[4m"

## Colors
COLOR_BLACK="\e[1m"
COLOR_RED="\e[1;31m"
COLOR_GREEN="\e[32m"
COLOR_DARK_YELLOW="\e[33m"
COLOR_MAGENTA="\e[1;35m"
COLOR_PURPLE="\e[35m"
COLOR_CYAN="\e[36m"
COLOR_BLUE="\e[1;34m"
COLOR_ORANGE="\e[31m"
COLOR_GREY="\e[37m"
COLOR_END="\e[0m"

# Prints an error message and exits.
function _printError()
{
	# Print the error text and exit
	string="${COLOR_RED}${STYLE_BOLD}ERROR: ${STYLE_RESET}${STYLE_BOLD}$1${STYLE_RESET} Exiting..."
	printf "%s\n" "$string"

	exit 1
}

# Checks if the current user is root.
function _checkRoot()
{
	if [ "$gID" -ne 0 ]; then
		# Re-run the script as root
		printf "This part of the script ${STYLE_UNDERLINED}needs${STYLE_RESET} to be run as root.\n"

		sudo clear
	fi
}

# Mounts the EFI system partition for the currently-booted disk.
function _mountEFI()
{
	# Find the BSD device name for the current OS disk
	osVolume=$(df / | awk '/disk/ {print $1}')

	# Find the EFI partition of the disk from the OS disk
	efiVolume=$(diskutil list "$osVolume" | awk '/EFI/ {print $6}')

	# Make sure the EFI partition actually exists
	if [ -z "$efiVolume" ]; then
		## Check if the OS is installed on a Core Storage (CS) logical volume
		csVolume=$(diskutil info "$osVolume" | grep "Core Storage")
		if [ ! -z "$csVolume" ]; then ## CS volume detected
			## We can find the recovery volume in the diskutil output, and then use that to find the EFI partition
			recoveryVolume=$(diskutil info "$osVolume" | awk '/Recovery Disk:/ {print $3}')
			efiVolume=$(diskutil list "$recoveryVolume" | awk '/EFI/ {print $6}')
		else ## No CS volume present, so assume no EFI partition
			printError "No EFI partition present on OS volume ($osVolume)!"
		fi
	fi

	# Check if the EFI partition is already mounted; if not, mount it
	if [ -z "$(mount | awk '/'"$efiVolume"'/ {print $1}')" ]; then
		diskutil mount "$efiVolume" > /dev/null
		gEFIMount=$(diskutil info "$efiVolume" | awk '/Mount Point/ {print $3}')
		echo "EFI system partition ($efiVolume) mounted at $gEFIMount."
	else
		gEFIMount=$(diskutil info "$efiVolume" | awk '/Mount Point/ {print $3}')
		echo "EFI system partition ($efiVolume) is already mounted at $gEFIMount."
	fi
}

# Installs a kext to the Extensions directory for the currently-booted disk.
function _installKext()
{
	# Fix the permissions of the kext
	sudo chmod -R 755 "$1"
	sudo chown -R 0:0 "$1"

	# Copy the kext to the proper extensions folder
	if [ $gOSXVersion -e 10.11 ]; then
		sudo mv "$1" /Library/Extensions
	else
		sudo mv "$1" /System/Library/Extensions
	fi
}

# Installs a kext to the Clover kexts directory for the currently-booted disk.
function _installKextEFI()
{
	# Copy the kext to the Clover kexts folder on the EFI partition
	cp -R "$1" "$gEFIMount/EFI/CLOVER/kexts/Other"
}

# Looks for a device in the IORegistry.
function _checkDevicePresence()
{
	ioreg -k compatible -p IODeviceTree | grep compatible | grep "pci$1,$2"
}

# Generates SMBIOS data for config.plist.
function _genSMBIOSData()
{
	# Initialize variables
	printf "${STYLE_BOLD}Generating SMBIOS data${STYLE_RESET}:\n"
	plist="$gEFIMount/EFI/CLOVER/config.plist"

	# Make sure config.plist exists
	if [ ! -f "$plist" ]; then
		_printError "config.plist not found!"
		exit 1
	fi

	serialNumber=$("$gRepo/modules/MacGen/simpleMacSerial.sh" $gProductName)
	MLB=$("$gRepo/modules/MacGen/simpleMLBSerial.sh" $gProductName $serialNumber)
	smUUID=$(uuidgen)

	echo " - Product Name: $gProductName"
	echo " - Serial Number: $serialNumber"
	echo " - MLB Serial Number: $MLB"
	echo " - System UUID: $smUUID"

	# Copy the generated data to the plist
	/usr/libexec/PlistBuddy -c "Set :SMBIOS:ProductName '$gProductName'" "$plist"
	/usr/libexec/PlistBuddy -c "Set :SMBIOS:SerialNumber '$serialNumber'" "$plist"
	/usr/libexec/PlistBuddy -c "Set :RtVariables:MLB '$MLB'" "$plist"
	/usr/libexec/PlistBuddy -c "Set :SMBIOS:SmUUID '$smUUID'" "$plist"
}


GA-Z77X.sh
#!/usr/bin/env zsh
#
# GA-Z77X.sh - Gigabyte GA-Z77X macOS Post-Install Script by theracermaster
# Gigabyte GA-Z77X DSDT Patch Repo - https://github.com/theracermaster/Gigabyte-GA-Z77X-DSDT-Patch

# Global variables
## The script version
gScriptVersion="2.0b1"

## The repo folder
gRepo=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)

## The motherboard, will be properly initialized later
gMotherboard="Unknown"
## The CPU microarchitecture, 2 = Sandy Bridge, 3 = Ivy Bridge
gCoreBridgeType=$(sysctl -n machdep.cpu.brand_string | cut -d '-' -f 2 | cut -c 1)
## The HDEF layout ID
gLayoutID=0
## PS/2 support, 0 = no, 1 = yes
gPS2=0

## The default Mac model (recommended: iMac13,1 or iMac13,2)
gProductName="iMac13,2"

source "$gRepo/tools/common.sh"

function _gitUpdate()
{
	cd "$gRepo"

	git pull

	# Update the submodules
	git submodule update --init --recursive
	git submodule foreach git pull origin master
}

function _identifyMotherboard()
{
	# Initialize variables
	motherboard=$(ioreg -k OEMBoard -p IODeviceTree | awk '/OEMBoard/ {print $4}' | tr -d '<"">')

	# Identify the motherboard
	case $motherboard in
		#'B75M-D3H')
		#	gMotherboard="B75M-D3H"
		#	gLayoutID=1
		#	;;
		'Z77X-D3H')
			gMotherboard="Z77X-D3H"
			gLayoutID=5
			gPS2=1
			;;
		'Z77X-UD3H')
			gMotherboard="Z77X-UD3H"
			gLayoutID=5
			gPS2=1
			;;
		#'Z77X-UD4H')
		#	gMotherboard="Z77X-UD4H"
		#	gLayoutID=1
		#	;;
		'Z77X-UD5H')
			gMotherboard="Z77X-UD5H"
			gLayoutID=1
			;;
		'Z77X-UP5')
			gMotherboard="Z77X-UP5-TH"
			gLayoutID=1
			;;
		#'Z77X-UP7')
		#	gMotherboard="Z77X-UP7"
		#	gLayoutID=1
		#	;;
		*)
			_printError "$motherboard is unsupported by this script!"
			;;
	esac
}

function _checkOSVersion()
{
	case $gOSVersion in
		10.11 | 10.12)
			;; # Supported OS version detected, so do nothing
		*)
			_printError "macOS Version $gOSVersion is unsupported by this script!";;
	esac
}

function _printHeader()
{
	# Initialize variables
	boardSeries=$(echo $gMotherboard | cut -d '-' -f 1)
	board=$(echo "${gMotherboard/$boardSeries-}")

	clear

	# Print the header & info
	echo "Gigabyte GA-Z77X.sh Post-Install Script v$gScriptVersion by theracermaster"
	echo "Updates & Info: https://github.com/theracermaster/Gigabyte-GA-Z77X-DSDT-Patch"
	echo "--------------------------------------------------------------------------------"
	printf "Detected motherboard: Gigabyte ${STYLE_BOLD}GA-$boardSeries-${COLOR_CYAN}$board${STYLE_RESET}\n"
	printf "Script arguments: ./GA-Z77X.sh $1\n"
	echo "--------------------------------------------------------------------------------"
}

function _detectAtherosNIC()
{
	# Gigabyte shipped different revisions of boards with different Atheros NICs
	# AtherosL1cEthernet supports AR8151, while AtherosE2200Ethernet supports AR8161
	# Install the correct Atheros kext if an Atheros NIC is detected
	if [ ! -z "$(_checkDevicePresence 1969 1083)" ]; then
 		# Atheros AR8151 v2.0 GbE - use AtherosL1cEthernet
		echo " - Atheros AR8151 v2.0 GbE [1969:1083] detected, installing AtherosL1cEthernet.kext..."
		_installKextEFI "$gRepo/drivers/kexts/AtherosL1cEthernet.kext"
	elif [ ! -z "$(_checkDevicePresence 1969 1091)" ]; then
		echo " - Atheros AR8161 GbE [1969:1091] detected, installing AtherosE2200Ethernet..."
		_installKextEFI "$gRepo/drivers/kexts/AtherosE2200Ethernet.kext"
	fi
}

function _detectIntelNIC()
{
	if [ ! -z "$(_checkDevicePresence 8086 1503)" ]; then
 		# Intel 82579V GbE - use IntelMausiEthernet
		echo " - Intel 82579V GbE [8086:1503] detected, installing IntelMausiEthernet..."
		_installKextEFI "$gRepo/drivers/kexts/IntelMausiEthernet.kext"
		case $gMotherboard in
			"Z77X-UD5H" | "Z77X-UP7") # Motherboards that have dual NICs
				/usr/libexec/PlistBuddy -c "Merge $gRepo/config/patches/acpi_DualLAN.plist ':ACPI:DSDT:Patches'" "$gEFIMount/EFI/CLOVER/config.plist"
				;;
			*) # Otherwise assume that the motherboard has a single Intel NIC
				/usr/libexec/PlistBuddy -c "Merge $gRepo/config/patches/acpi_SingleLAN.plist ':ACPI:DSDT:Patches'" "$gEFIMount/EFI/CLOVER/config.plist"
				;;
		esac
	fi
}

function _detectRealtekNIC()
{
	if [ ! -z "$(_checkDevicePresence 10ec 8168)" ]; then
 		# Realtek RTL8168/RTL8111 GbE - use RealtekRTL8111
		echo " - Realtek RTL8168/RTL8111 GbE [10EC:8168] detected, installing RealtekRTL8111..."
		_installKextEFI "$gRepo/drivers/kexts/RealtekRTL8111.kext"
	fi
}

function _detectMarvellSATA()
{
	if [ ! -z "$(_checkDevicePresence 1b4b 9172)" ]; then
 		# Marvell 88SE9172 SATA - use GenericAHCIPort
		echo " - Marvell 88SE9172 SATA [1B4B:9172] detected, installing GenericAHCIPort..."
		_installKextEFI "$gRepo/drivers/kexts/GenericAHCIPort.kext"
	fi
}

function _detectXHCI()
{
	# Add AppleUSBXHCIPCI injector for non-Intel xHCI controllers
	if [ ! -z "$(_checkDevicePresence 1106 3432)" ]; then
		echo " - VIA VL800 xHCI [1106:3432] detected, installing GenericUSBXHCIPCI.kext..."
		_installKextEFI "$gRepo/drivers/kexts/GenericUSBXHCIPCI.kext"
	#elif [ ! -z "$(_checkDevicePresence 1B6F 7023)" ]; then
	#	echo " - Etron EJ168 xHCI [1B6F:7023] detected, installing GenericUSBXHCIPCI.kext..."
	#	_installKextEFI "$gRepo/drivers/kexts/GenericUSBXHCIPCI.kext"
	fi
}

function _detectFireWire()
{
	# Initialize variables
	plist="$gEFIMount/EFI/CLOVER/config.plist"

	# Add EFI driver to fix GUID issues with built-in FireWire
	## Add the kext patches to config.plist if non-Intel xHCI controllers (VIA VL810/Etron EJ168) are present
	if [ ! -z "$(_checkDevicePresence 1106 3044)" ]; then
		echo " - VIA VT6308 FireWire OHCI [1106:3044] detected, installing EFI drivers..."
		cp "$gRepo/drivers/efi/FireWireDevice.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/FireWireDevice.efi"
		cp "$gRepo/drivers/efi/FireWireOhci.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/FireWireOhci.efi"
	fi
}

function _detectThunderbolt()
{
	# Add SSDT to (potentially) fix some Thunderbolt issues
	if [ ! -z "$(_checkDevicePresence 8086 1547)" ]; then
		echo " - Compiling SSDT-TB.dsl..."
		"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-TB.aml" "$gRepo/acpi/common/SSDT-TB.dsl" |& tail -n 1
	fi

	echo
}

function _detectPS2()
{
	if [ $gPS2 -eq 1 ]; then
		echo " - PS/2 hardware present, installing VoodooPS2Controller..."
		_installKextEFI "$gRepo/drivers/kexts/VoodooPS2Controller.kext"
	fi
}

function _compileACPI()
{
	# Initialize variables
	ssdtCFG="SSDT-CFG-$boardSeries-$board.dsl"

	# Compile ACPI source files and move output files to the ESP
	echo " - Compiling $ssdtCFG:"
	"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-CFG.aml" "$gRepo/acpi/$ssdtCFG" |& tail -n 1
	echo " - Compiling SSDT-HDA.dsl:"
	"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-HDA.aml" "$gRepo/acpi/common/SSDT-HDA.dsl" |& tail -n 1
	echo " - Compiling SSDT-IGPU.dsl:"
	"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-IGPU.aml" "$gRepo/acpi/common/SSDT-IGPU.dsl" |& tail -n 1
	echo " - Compiling SSDT-PEG.dsl:"
	"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-PEG.aml" "$gRepo/acpi/common/SSDT-PEG.dsl" |& tail -n 1
	echo " - Compiling SSDT-USB.dsl:"
	"$IASL" -p "$gEFIMount/EFI/CLOVER/ACPI/patched/SSDT-USB.aml" "$gRepo/acpi/common/SSDT-USB.dsl" |& tail -n 1
}

function _install()
{
	# Mount the EFI system partition
	_mountEFI

	# Clear the output and print the header
	_printHeader "${STYLE_BOLD}--install: ${COLOR_GREEN}Installing Clover Bootloader${STYLE_RESET}"

	# Check if there is an existing bootloader install; if so, ask the user if it can be overwritten
	if [ -d "$gEFIMount/EFI/CLOVER" ]; then
		echo "An existing installation of Clover UEFI was found on the EFI system partition."
		printf "${STYLE_BOLD}Do you want to overwrite it (y/n)?${STYLE_RESET} "
		read choice
		case "$choice" in
			y|Y)
				echo "Removing existing Clover UEFI bootloader install..."
				rm -rf "$gEFIMount/EFI";;
			n|N)
				echo "Backing up existing Clover UEFI bootloader install..."
				mv "$gEFIMount/EFI/BOOT" "$gEFIMount/BOOTbackup"
				mv "$gEFIMount/EFI/CLOVER" "$gEFIMount/CLOVERbackup";;
		esac
	fi

	# Clear the output and reprint the header
	_printHeader "${STYLE_BOLD}--install: ${COLOR_GREEN}Installing Clover Bootloader${STYLE_RESET}"

	# Copy Clover files/folders
	mkdir -p "$gEFIMount/EFI"
	cp -R "$gRepo/efi/BOOT" "$gEFIMount/EFI"
	cp -R "$gRepo/efi/CLOVER" "$gEFIMount/EFI"
	cp "$gRepo/config/config_main.plist" "$gEFIMount/EFI/CLOVER/config.plist"

	printf "${STYLE_BOLD}Installing required kexts${STYLE_RESET}:\n"
	# Install the required kexts
	_installKextEFI "$gRepo/drivers/kexts/FakeSMC.kext"
	_installKextEFI "$gRepo/drivers/kexts/AppleALC.kext"
	_installKextEFI "$gRepo/drivers/kexts/Shiki.kext"

	# Install kexts for other detected hardware
	_detectAtherosNIC
	_detectIntelNIC
	_detectRealtekNIC
	_detectMarvellSATA
	_detectXHCI
	_detectPS2

	# Set layout ID in config.plis
	/usr/libexec/PlistBuddy -c "Set ':Devices:Audio:Inject' $gLayoutID" "$gEFIMount/EFI/CLOVER/config.plist"

	# Install the required UEFI drivers
	cp "$gRepo/drivers/efi/EfiDevicePathPropertyDatabase.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/EfiDevicePathPropertyDatabase.efi"
	cp "$gRepo/drivers/efi/FSInject.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/FSInject.efi"
	cp "$gRepo/drivers/efi/HfsPlus.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/HfsPlus.efi"
	cp "$gRepo/drivers/efi/OsxAptioFix2Drv.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/OsxAptioFix2Drv.efi"
	cp "$gRepo/drivers/efi/PartitionDxe.efi" "$gEFIMount/EFI/CLOVER/drivers64uefi/PartitionDxe.efi"

	# Install the FireWire EFI drivers (if needed)
	_detectFireWire

	# Generate the SMBIOS data
	_genSMBIOSData

	# Clear the output and reprint the header
	printf "\n${STYLE_BOLD}Press enter to continue...${STYLE_RESET}\n" && read
	_printHeader "${STYLE_BOLD}--install: ${COLOR_BLUE}Compiling ACPI Source Files${STYLE_RESET}"

	# Compile ACPI source files
	printf "${STYLE_BOLD}Compiling SSDTs${STYLE_RESET}:\n"
	_compileACPI
	_detectThunderbolt

	# We're done here, let's prompt the user to reboot
	printf "${STYLE_BOLD}Installation complete. Do you want to reboot now (y/n)?${STYLE_RESET} " && read choice
	case "$choice" in
		y|Y) # User said yes, so let's reboot
			echo "Rebooting..."
			sudo reboot
			;;
		*) # Anything else, let's just quit
			echo "Exiting..."
			exit 0
			;;
	esac

}

_identifyMotherboard
_checkOSVersion

case "$1" in
#	--git-update)
#		gitUpdate
#		;;
	--install)
		_gitUpdate
		_install
		;;
	*)
		echo "Gigabyte GA-Z77X.sh Post-Install Script v$gScriptVersion by theracermaster"
		echo "Supports various Gigabye GA-Z77X motherboards running macOS 10.10+"
		echo
		echo "Usage: ./GA-Z77X.sh <command>, where <command> is one of the following:"
		echo
		echo "     --git-update         Update the repo to the latest version"
		echo "     --install            Install Clover UEFI to your EFI system partition"
		echo
		echo "Updates & Info: https://github.com/theracermaster/Gigabyte-GA-Z77X-DSDT-Patch"
esac

exit 0
